app/src/combo.c

```diff
@@ -105,16 +107,37 @@ static int initialize_combo(struct combo_cfg *new_combo) {
 }
 
 static int setup_candidates_for_first_keypress(int32_t position, int64_t timestamp) {
+    // Due to layer filtering we need to track the candidates[index] stuff by hand
+    int number_of_combo_candidates = -1; // First increment moves this to 0 as expected
+    uint8_t highest_active_layer = zmk_keymap_highest_layer_active();
     for (int i = 0; i < CONFIG_ZMK_COMBO_MAX_COMBOS_PER_KEY; i++) {
+        number_of_combo_candidates++; // Increment on each pass of the loop
         struct combo_cfg *combo = combo_lookup[position][i];
         if (combo == NULL) {
-            return i;
+            return number_of_combo_candidates;
+        }
+        // Filter combo by layer
+        // -1 in the first layer position is global layer scope for the combo (aka valid)
+        bool combo_layer_valid = false;
+        if (combo->layers[0] == -1) {
+            combo_layer_valid = true;
+        } else {
+            for (int j = 0; j < sizeof(combo->layers) / sizeof(combo->layers[0]); j++) {
+                if (combo->layers[j] == highest_active_layer) {
+                    combo_layer_valid = true;
+                    break;
+                }
+            }
+        }
```

Comment:
I think we should extract lines 121-131 into a separate function like `combo_active_on_layer(combo, layer)`

Reply:
Extracted this as a method per suggestion